core/test/Cardano/Address/ScriptSpec.hs

```diff
@@ -296,28 +331,158 @@ spec = do
+                                   ]
+                    ])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left ListTooSmall)
+
+        it "duplicate content in RequireAllOf" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireAllOf [cosigner1, cosigner2, cosigner1])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left DuplicateSignatures)
+
+        it "duplicate content in RequireAnyOf" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireAnyOf [cosigner1, cosigner2, cosigner1])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left DuplicateSignatures)
+
+        it "duplicate content in RequireSomeOf" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1 [cosigner1, cosigner2, cosigner1])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left DuplicateSignatures)
+
+        it "duplicate in nested" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1
+                    [ cosigner1
+                    , RequireAnyOf [ cosigner2
+                                   , RequireSomeOf 2 [cosigner0, cosigner0, cosigner3]
+                                   ]
+                    ])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left DuplicateSignatures)
+
+        it "invalid timelocks - too many" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1 [cosigner0, ActiveFromSlot 1, ActiveFromSlot 2, ActiveUntilSlot 20])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left InvalidTimelocks)
+
+        it "invalid timelocks - contradictory 1" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1 [cosigner0, ActiveFromSlot 21, ActiveUntilSlot 20])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left InvalidTimelocks)
+
+        it "invalid timelocks - contradictory 1" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 1
+                    [ cosigner1
+                    , RequireAnyOf [ cosigner2
+                                   , RequireSomeOf 2 [cosigner0, cosigner1, ActiveFromSlot 21, ActiveUntilSlot 20, cosigner3]
+                                   ]
+                    ])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left InvalidTimelocks)
+
+        it "too high m in RequireSomeOf when timelocks" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSomeOf 3 [cosigner0, cosigner1, ActiveFromSlot 21, ActiveUntilSlot 30])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left ListTooSmall)
+
+        it "no content in RequireAnyOf when timelocks" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireAnyOf [ActiveFromSlot 21, ActiveUntilSlot 30])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left EmptyList)
```

Comment:
I don't see why this is an invalid script? Or, is it something actually forbidden by the ledger? I'd suggest to revise the error type then if it is because `EmptyList` is a bit confusing, the list isn't empty but rather lacks of required key hashes (if they are indeed required?). I don't recall this however :thinking: ?

Reply:
not ledger (required) and recommended validation is present. There are numerous tests that show what is required, what is recommended (aka I found reasonable)